guide/src/trait_bounds.md

```diff
@@ -0,0 +1,509 @@
+
+What happens if you have type errors when using Python and how can you improve the error messages?
+
+
+### Wrong types in Python function arguments
+
+Let's assume in the first case that you will use in your Python file `my_rust_model.set_variables_in_rust(2.0)` instead of `my_rust_model.set_variables_in_rust([2.0])`.
+
+The Rust signature expects a vector, which corresponds to a list in Python.
+What happens if instead of a vector, we pass a single value ?
+
+At the execution of Python, we get :
+
+```block
+File "main.py", line 15, in <module>
+   my_rust_model.set_variables_in_rust(2)
+TypeError
+```
+
+It is a type error and Python points to it, so it's easy to identify and solve.
+
+### Wrong types in Python method signatures
+
+Let's assume now that the return type of one of the methods of our Model class is wrong, for example the `get_results` method that is expected to return a `Vec<f64>` in Rust, a list in Python.
+
+```python
+class Model:
+    def set_variables(self, inputs):
+        self.inputs = inputs
+    def compute(self):
+        self.results = [elt**2 -3 for elt in self.inputs]
+    def get_results(self):
+        return self.results[0]
+        #return self.results <-- this is the expected output
+```
+
+This call results in the following panic:
+
+```block
+pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: PyErr { type: Py(0x10dcf79f0, PhantomData) }
+```
+
+This error code is not helpful for a Python user that does not know anything about Rust, or someone that does not know PyO3 was used to interface the Rust code.
+
+However, as we are responsible for making the Rust code available to Python, we can do something about it.
+
+The issue is that we called `unwrap` anywhere we could, and therefore any panic from PyO3 will be directly forwarded to the end user.
+
+Let's modify the code performing the type conversion to give a helpful error message to the Python user:
+
+We used in our `get_results` method the following call that performs the type conversion:
+
+```rust
+impl Model for UserModel {
+  fn get_results(&self) -> Vec<f64> {
+    println!("Get results from Rust calling Python");
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    self
+        .model
+        .as_ref(py)
+        .call_method("get_results", (), None)
+        .unwrap()
+        .extract()
+        .unwrap()
+}
+```
+
+Let's break it down in order to perform better error handling:
+
+```rust
+impl Model for UserModel {
+  fn get_results(&self) -> Vec<f64> {
+      println!("Get results from Rust calling Python");
+      let gil = Python::acquire_gil();
+      let py = gil.python();
+      let py_result: &PyAny = self
+          .model
+          .as_ref(py)
+          .call_method("get_results", (), None)
+          .unwrap();
+
+      if py_result.get_type().name() != "list" {
+          panic!("Expected a list for the get_results() method signature, got {}", py_result.get_type().name());
```

Comment:
Instead of panicking, the UserModel trait could be modified to return some form of `Result`. Did you consider this option, and if so, why did you choose against it?

Reply:
I did not consider it as I assumed the trait could be changed as I was in a scenario of having a code base that we wanted to make available in Python without actually changing it.